torch/csrc/jit/passes/onnx/fold_if_node.cpp

```diff
@@ -0,0 +1,229 @@
+    auto res = at::less(inputs[0], inputs[1]);
+    if (prev_node->kind() == onnx::Not)
+      res = at::less_equal(inputs[0], inputs[1]);
+    return at::is_nonzero(res);
+  }
+  return false;
+}
+
+// before peephole pass
+// graph(%y.2 : Int(3, 4, strides=[4, 1], requires_grad=0, device=cpu)):
+//   %4 : Long(2, strides=[1], device=cpu) = onnx::Shape(%y.2)
+//   %5 : Long(device=cpu) = onnx::Size(%4)
+//   %12 : Long(requires_grad=0, device=cpu) = onnx::Constant[value={2}]()
+//   %6 : Bool(device=cpu) = onnx::Equal(%5, %12)
+//   %11 : bool = onnx::Cast[to=9](%6)
+//   %7 : Int(3, 4, strides=[4, 1], device=cpu) = onnx::If(%11)
+//     block0():
+//       %13 : Int(requires_grad=0, device=cpu) = onnx::Constant[value={4}]()
+//       %8 : Int(3, 4, strides=[4, 1], device=cpu) = onnx::Add(%y.2, %13)
+//       %14 : Int(requires_grad=0, device=cpu) = onnx::Constant[value={2}]()
+//       %9 : Int(3, 4, strides=[4, 1], device=cpu) = onnx::Add(%8, %14)
+//       -> (%9)
+//     block1():
+//       %y.1 : Int(3, 4, strides=[4, 1], requires_grad=0, device=cpu) =
+//       onnx::Identity(%y.2)
+//       -> (%y.1)
+//   return (%7)
+
+// after peephole pass
+// graph(%y.2 : Int(3, 4, strides=[4, 1], requires_grad=0, device=cpu)):
+//   %4 : Long(2, strides=[1], device=cpu) = onnx::Shape(%y.2)
+//   %5 : Long(device=cpu) = onnx::Size(%4)
+//   %12 : Long(requires_grad=0, device=cpu) = onnx::Constant[value={2}]()
+//   %6 : Bool(device=cpu) = onnx::Equal(%5, %12)
+//   %11 : bool = onnx::Cast[to=9](%6)
+//   %13 : Int(requires_grad=0, device=cpu) = onnx::Constant[value={4}]()
+//   %8 : Int(3, 4, strides=[4, 1], device=cpu) = onnx::Add(%y.2, %13)
+//   %14 : Int(requires_grad=0, device=cpu) = onnx::Constant[value={2}]()
+//   %9 : Int(3, 4, strides=[4, 1], device=cpu) = onnx::Add(%8, %14)
+//   return (%9)
+
+static void foldIfNode(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      foldIfNode(child_block);
+    }
+    if (it->kind() == onnx::If) {
+      auto if_node = *it;
+      if (conditionIfFold(if_node)) {
+        Block* then_block = it->blocks()[0];
+        Block* else_block = it->blocks()[1];
+        Block* block = else_block;
+        if (constantFoldingValue(if_node))
+          block = then_block;
+
+        std::vector<Node*> nodes_in_valid_path;
+        for (auto* valid_node : block->nodes()) {
+          nodes_in_valid_path.push_back(valid_node);
+        }
+        Node* cur = if_node;
+        for (auto* valid_node : nodes_in_valid_path) {
+          valid_node->moveAfter(cur);
+          cur = valid_node;
+        }
+        for (size_t i = 0; i < block->return_node()->inputs().size(); ++i) {
+          if_node->outputs()[i]->replaceAllUsesWith(
+              block->return_node()->inputs()[i]);
+        }
+        it->removeAllInputs();
+        it.destroyCurrent();
+      }
+    }
+  }
+}
+
+void FoldIfONNX(Block* b) {
```

Comment:
Please add a few lines of comments explaining what each function is doing, and how to use/when to use them

Reply:
I have added comments